File: flow-client/src/main/java/com/vaadin/client/communication/XhrConnectionError.java
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.vaadin.client.communication;

import jsinterop.annotations.JsType;

import com.google.gwt.xhr.client.XMLHttpRequest;

import elemental.json.JsonObject;

/**
 * Detail about an error which occurred during an XHR request to the server.
 * Pure {@code @JsType(isNative=true)} binding to the TypeScript implementation
 * at
 * {@code src/main/frontend/internal/client/communication/XhrConnectionError.ts}.
 *
 * @author Vaadin Ltd
 * @since 1.0
 */
@JsType(isNative = true, namespace = "Vaadin.Flow.internal.client.communication", name = "XhrConnectionError")
public class XhrConnectionError {

    public XhrConnectionError(XMLHttpRequest xhr, JsonObject payload,
            Exception exception) {
        // Defined by the TS class constructor.
    }

    /**
     * Returns the exception which caused the problem, if available.
     */
    public native Exception getException();

    /**
     * Returns the {@link XMLHttpRequest} which failed to reach the server.
     */
    public native XMLHttpRequest getXhr();

    /**
     * Returns the payload which was sent to the server.
     */
    public native JsonObject getPayload();
}

File: flow-client/src/main/frontend/internal/communication/XhrConnectionError.ts
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

// TypeScript port of com.vaadin.client.communication.XhrConnectionError, built
// alongside the Java version. A data holder describing an error during an XHR
// request to the server.

/** Detail about an error during an XHR request to the server; mirrors XhrConnectionError.java. */
export class XhrConnectionError {
  private readonly xhr: XMLHttpRequest;

  private readonly payload: Record<string, unknown>;

  private readonly error: Error | null;

  constructor(xhr: XMLHttpRequest, payload: Record<string, unknown>, error: Error | null) {
    this.xhr = xhr;
    this.payload = payload;
    this.error = error;
  }

  /** The error that caused the problem, or null if not available. */
  getException(): Error | null {
    return this.error;
  }

  /** The request that failed to reach the server. */
  getXhr(): XMLHttpRequest {
    return this.xhr;
  }

  /** The payload that was sent to the server, never null. */
  getPayload(): Record<string, unknown> {
    return this.payload;
  }
}
